# Homework 2 Assignment

## Task

Using previously downloaded dataset, conduct the following steps:

1. Create a `DataFrame` using the `pd.DataFrame` command;
2. Review the data using the commands `shape`, `info()`, `describe()`, `columns`, `head()`;
3. Select target data using `drop()`, select .columns "if" condition;
4. View and delete duplicate data using the `duplicated()`, `drop_duplicates()`;
5. View empty data and delete using the `isna()`, `dropna()`;

## Solution

### Task 1. Loading the dataframe

Similarly to the previous task, we download the dataset

In [3]:
import pandas as pd

# Specifying path to the file containing our dataset
DATASET_PATH = 'StudentsPerformance.csv'

# Loading the dataframe
df = pd.read_csv(DATASET_PATH, delimiter=',')
df.head(6) # Showing first 6 entries

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78

### Task 2. Inspecting data

Now, checking info about the selected dataset

In [14]:
# Showing shape
print(f'Shape of the dataset is {df.shape}')

# Showing information
print('\n--- Information about the dataset ---')
df.info()

# Showing columns
print('\n--- Dataset columns ---')
print([column for column in df.columns])

# Dataset description
print('\n--- Dataset description ---')
df.describe()

Shape of the dataset is (1000, 8)

--- Information about the dataset ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB

--- Dataset columns ---
['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

--- Dataset description ---


math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

### Task 3. Dropping and selecting

Let us drop "race/ethnicity" and "lunch" columns as they are probably not relevant to the resultant scores

In [15]:
df = df.drop(['race/ethnicity', 'lunch'], axis=1)
df

gender parental level of education test preparation course  math score  \
0    female           bachelor's degree                    none          72   
1    female                some college               completed          69   
2    female             master's degree                    none          90   
3      male          associate's degree                    none          47   
4      male                some college                    none          76   
..      ...                         ...                     ...         ...   
995  female             master's degree               completed          88   
996    male                 high school                    none          62   
997  female                 high school               completed          59   
998  female                some college               completed          68   
999  female                some college                    none          77   

     reading score  writing score  
0               72             74  
1               90             88  
2               95             93  
3               57             44  
4               78             75  
..             ...            ...  
995             99             95  
996             55             55  
997             71             65  
998             78             77  
999             86             86  

[1000 rows x 6 columns]

From remaining columns let us select everything except for "gender"

In [19]:
selected_columns = [column for column in df.columns if column != 'gender']

print(f'Remaining columns except for "gender" are:\n{selected_columns}')

Remaining columns except for "gender" are:
['parental level of education', 'test preparation course', 'math score', 'reading score', 'writing score']


### Step 4. Duplicates

Let us inspect and find some duplicates in data

In [49]:
duplicated = df.duplicated()
duplicated

0      False
1      False
2      False
3      False
4      False
       ...  
995    False
996    False
997    False
998    False
999    False
Length: 1000, dtype: bool

It is unclear whether the data has duplicates at this point, so let us print the number of duplicates

In [56]:
duplicates_number = len([duplicated_true for duplicated_true in duplicated if duplicated_true is True])
print(f"The duplicates number is {duplicates_number}")

The duplicates number is 2


Aha! So we do have some duplicates! Let us print them

In [76]:
df[df.duplicated(keep=False) == True]

gender parental level of education test preparation course  math score  \
452  female          associate's degree                    none          65   
613  female          associate's degree                    none          65   
632  female           bachelor's degree               completed          66   
692  female           bachelor's degree               completed          66   

     reading score  writing score  
452             77             74  
613             77             74  
632             74             81  
692             74             81

So we have only two duplicates. However, let us add some more!

Select first 10 columns and append to the dataframe

In [74]:
DUPLICATES_NUMBER = 10 # Define the number of duplicates to add

df_duplicate_rows = df[:DUPLICATES_NUMBER] # We are going to add these rows to the existing dataframe
corrupted_df = pd.concat([df, df_duplicate_rows], ignore_index=True)

# Asserting that we indeed added the proper number of duplicates
assert corrupted_df.shape[0] == df.shape[0] + DUPLICATES_NUMBER, 'duplicates added incorrectly!'

# Now, let us see what .duplicated() gives us
corrupted_df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1005     True
1006     True
1007     True
1008     True
1009     True
Length: 1010, dtype: bool

Let us drop duplicates now:

In [86]:
df = corrupted_df.drop_duplicates()
df

gender parental level of education test preparation course  math score  \
0    female           bachelor's degree                    none          72   
1    female                some college               completed          69   
2    female             master's degree                    none          90   
3      male          associate's degree                    none          47   
4      male                some college                    none          76   
..      ...                         ...                     ...         ...   
995  female             master's degree               completed          88   
996    male                 high school                    none          62   
997  female                 high school               completed          59   
998  female                some college               completed          68   
999  female                some college                    none          77   

     reading score  writing score  
0               72             74  
1               90             88  
2               95             93  
3               57             44  
4               78             75  
..             ...            ...  
995             99             95  
996             55             55  
997             71             65  
998             78             77  
999             86             86  

[998 rows x 6 columns]

As expected, since we had 12 duplicates, we now have 998 rows!

### Step 5. Empty data

Let us first check whether there are any rows with empty values:

In [91]:
print(
    'dataset contains no empty values' 
    if len(df[df.isna().any(axis=1)]) == 0 
    else 'dataset contains some empty values'
)

dataset contains no empty values


Yikes, our dataset is too perfect. Let us add some empty values

In [96]:
# Creating corrupted data with empty rows
df_empty_rows = pd.DataFrame(data = {
    'gender': ['female', 'male', 'female'],
    'parental level of education': ['high school', 'some college', 'high school'],
    'test preparation course': ['none', 'none', 'completed'],
    'math score': [100, 50, 100],
    'reading score': [None, 50, 100],
    'writing score': [100, 50, None]
})
assert len(df_empty_rows[df_empty_rows.isna().any(axis=1)]) == 2, 'we should have two additional corrupted rows!'

# Concetenating our dataframe with the corrupted one
corrupted_df = pd.concat([df, df_empty_rows], ignore_index=True)

corrupted_df.tail(5) # Showing last five lines with three of them being new

gender parental level of education test preparation course  math score  \
996   female                some college               completed          68   
997   female                some college                    none          77   
998   female                 high school                    none         100   
999     male                some college                    none          50   
1000  female                 high school               completed         100   

      reading score  writing score  
996            78.0           77.0  
997            86.0           86.0  
998             NaN          100.0  
999            50.0           50.0  
1000          100.0            NaN

Now, debugging the empty elements

In [99]:
corrupted_df[corrupted_df.isna().any(axis=1)]

gender parental level of education test preparation course  math score  \
998   female                 high school                    none         100   
1000  female                 high school               completed         100   

      reading score  writing score  
998             NaN          100.0  
1000          100.0            NaN

Removing corrupted rows!

In [100]:
corrupted_df.dropna()

gender parental level of education test preparation course  math score  \
0    female           bachelor's degree                    none          72   
1    female                some college               completed          69   
2    female             master's degree                    none          90   
3      male          associate's degree                    none          47   
4      male                some college                    none          76   
..      ...                         ...                     ...         ...   
994    male                 high school                    none          62   
995  female                 high school               completed          59   
996  female                some college               completed          68   
997  female                some college                    none          77   
999    male                some college                    none          50   

     reading score  writing score  
0             72.0           74.0  
1             90.0           88.0  
2             95.0           93.0  
3             57.0           44.0  
4             78.0           75.0  
..             ...            ...  
994           55.0           55.0  
995           71.0           65.0  
996           78.0           77.0  
997           86.0           86.0  
999           50.0           50.0  

[999 rows x 6 columns]

So now we have indeed by 2 rows less!